### DISCUSSION TITLE
About Processor simulation

### DISCUSSION THREAD

**Discussion by mmcheraghi** [stnolting/neorv32#967] [Q&A]
Hi @stnolting 
How are you?

Sorry I have got stuck at sth. There are two versions of a program each with an specific custom instruction. I use hardware performance counters to measure elapsed cycles.

The superior method in clock cycles **takes longer** in simulation. In other words, it takes fewer cycles according to cycle timer but it finishes simulation some time later than the other one. What may cause this?

Thanks for your attention.

**Comment by mmcheraghi** [comment]
This is my top file and configurations.
[neorv32_top.zip](https://github.com/user-attachments/files/16405498/neorv32_top.zip)

**Comment by stnolting** [comment]
Hey @mahdi259.

I'm sorry, but I don't think I understand the problem. 🙈

> The superior method in clock cycles takes longer in simulation.

Superior in terms of... what? Clock cycles?
So, it requires **less** clock cycles (according to the CPU timers/counters) but takes **longer** to simulate?

**Reply by mmcheraghi** [reply]
Yes, superior in terms of clock cycles. Yes

**Reply by stnolting** [reply]
Got it!

I am not really sure what is going on without inspecting the actual program code. But here is one theory... 🤔

Your program A requires **less clock cycles** but takes longer to simulate. In contrast, program B requires **more clock cycles** but the simulation is faster. Correct?

So program A is performing additional work (= simulation takes longer) but that additional work is not captured by the performance counters. So this additional effort must be in the `crt0` start-up code, which is executed before the actual `main()`.

Among other things, crt0 is responsible for intializing the `.data` segment and clearing the `.bss` segment. So if program A uses a lot of global variables there is additional work to do in crt0:

#### `.data`

Global variable that are **NOT initialized at declaration** are placed into the `.data` segment. For example:

```
int global_var_x = 0;
```

crt0 will initialize these variables from the `rodata` (read-only data) section: 

https://github.com/stnolting/neorv32/blob/cfff523cb76d66e42b3fb00fa00aec98ca41b1b8/sw/common/crt0.S#L85-L102

#### `.bss`

Global variable that are **NOT initialized at declaration** are placed into the `.bss` segment. For example:

```
int global_var_y;
```

crt0 will clear the entire segment initializing all those variables with zero:

https://github.com/stnolting/neorv32/blob/cfff523cb76d66e42b3fb00fa00aec98ca41b1b8/sw/common/crt0.S#L105-L118

----------------------------------------

Since you are starting benchmarking (reset performance counters) at the beginning of `main()` (when crt0 has already completed execution) this additional work in not captured by the obtained clock cycles. But you can "see" the additional workload as increased simulation time.

**Reply by mmcheraghi** [reply]
Thanks a lot.

**Reply by stnolting** [reply]
You're welcome!

The performance counters are cleared by a hardware reset. So if you remove your manual counter reset at the beginning of your programs you should see the the crt0-caused workload as well (at lest in the cycle counter).
